specs/0000-st-app/product-spec.md

```diff
@@ -0,0 +1,419 @@
+  Streamlit (requires custom endpoint support)
+- [#8713](https://github.com/streamlit/streamlit/issues/8713) — Custom 404/Error/Maintenance
+  pages
+
+**Security Headers & Middleware:**
+
+- [#6417](https://github.com/streamlit/streamlit/issues/6417) — Handling security headers
+  (CSP, HSTS, X-Frame-Options) — **34+ upvotes**
+- [#9160](https://github.com/streamlit/streamlit/issues/9160) — Allow configuring HTTP
+  security headers
+- [#861](https://github.com/streamlit/streamlit/issues/861) — Ability to write cookies for
+  auth persistence
+- [#8823](https://github.com/streamlit/streamlit/issues/8823) — IP whitelist for access
+  control
+
+**Framework Integration & ASGI Support:**
+
+- [#4311](https://github.com/streamlit/streamlit/issues/4311) — Run with WSGI/ASGI
+  protocols (deploy alongside Django in same container) — **67+ upvotes**
+- [#4567](https://github.com/streamlit/streamlit/issues/4567) — Integration with
+  Flask/FastAPI — **83+ upvotes**
+- [#927](https://github.com/streamlit/streamlit/issues/927) — Run on Django via Django
+  Channels
+- [#8661](https://github.com/streamlit/streamlit/issues/8661) — Expose Tornado instance
+  for low-level access
+- [#7546](https://github.com/streamlit/streamlit/issues/7546) — Multiple Runtime instances
+  in single process
+
+**Server Configuration & Lifecycle:**
+
+- [#7688](https://github.com/streamlit/streamlit/issues/7688) — On session start/shutdown
+  hooks for global resources
+- [#9916](https://github.com/streamlit/streamlit/issues/9916) — Tornado HTTPServer extra
+  arguments (max_header_size, max_body_size)
+- [#8991](https://github.com/streamlit/streamlit/issues/8991) — Server startup hooks for
+  cache pre-warming
+- [#6108](https://github.com/streamlit/streamlit/issues/6108) — Pre-cache on start-up
+  (populate cache before first user connects)
+- [#8545](https://github.com/streamlit/streamlit/issues/8545) — Detect browser tab close
+  for session cleanup
+- [#7667](https://github.com/streamlit/streamlit/issues/7667) — Setup script entry point
+  for initialization outside script runner
+
+**SEO & Metadata:**
+
+- [#5673](https://github.com/streamlit/streamlit/issues/5673) — Custom website metadata
+  (manifest, icons, description)
+- [#8999](https://github.com/streamlit/streamlit/issues/8999) — Embed metadata for SEO
+  indexing (Google Analytics, Search Console)
+
+## Proposal
+
+### API
+
+The `st.App` signature follows the [Starlette Application](https://www.starlette.io/applications/) signature, with
+`script_path` replacing Starlette's `routes` as the primary parameter:
+
+```python
+st.App(
+    script_path: str | Path,
+    *,
+    lifespan: Callable[[App], AsyncContextManager[dict[str, Any] | None]] | None = None,
+    routes: Sequence[BaseRoute] | None = None,
+    middleware: Sequence[Middleware] | None = None,
+    exception_handlers: Mapping[Any, ExceptionHandler] | None = None,
+    debug: bool = False,
```

Comment:
Should we explicitly add these parameters here or just have `**kwargs` that we pass on to the Starlette object? If we list the params explicitly, what happens if Starlette changes something? Or is that unlikely since it's such a fundamental API?